multichain-api-gateway/src/lib.rs

```diff
@@ -0,0 +1,211 @@
+    web,
+    web::{Data, Json},
+    App, HttpRequest, HttpResponse, HttpServer, Responder,
+};
+use futures::{stream, StreamExt};
+use reqwest::Client;
+use serde_json::Value;
+use url::Url;
+
+use crate::config::{BlockScoutSettings, Instance, Settings};
+
+mod cli;
+pub mod config;
+
+#[derive(Clone)]
+pub struct APIsEndpoints {
+    apis: Vec<(Instance, Url)>,
+    concurrent_requests: usize,
+}
+
+/// Assumptions:
+/// 1. All api calls expected to have trailing path "/api"
+///     e.g. "<base_url>/<...>/<...>/api?<query>"
+/// 2. First two segments of path of api call expected to be (network, chain)
+///     e.g. "<base_url>/<network>/<chain>/<...>/<...>"
+/// Taking it to account, we expect the following api call urls:
+///     e.g. <base_url>/<network>/<chain>/api?<query>   
+impl TryFrom<BlockScoutSettings> for APIsEndpoints {
+    type Error = &'static str;
+
+    fn try_from(settings: BlockScoutSettings) -> Result<Self, Self::Error> {
+        let mut apis = Vec::new();
+        for Instance(net, subnet) in settings.instances {
+            let mut url = settings.base_url.clone();
+            url.set_path(&format!("{}/{}/api", net, subnet));
+            apis.push((Instance(net, subnet), url));
+        }
+        Ok(Self {
+            apis,
+            concurrent_requests: settings.concurrent_requests,
+        })
+    }
+}
+
+async fn make_requests(apis_endpoints: APIsEndpoints) -> Vec<(Instance, String)> {
+    let client = Client::new();
+
+    stream::iter(apis_endpoints.apis)
+        .map(|(instance, url)| async {
+            let resp = client.get(url).send().await.unwrap();
+            (instance, resp.bytes().await)
+        })
+        .buffer_unordered(apis_endpoints.concurrent_requests)
+        .map(|(instance, response)| match response {
+            Ok(bytes) => (
+                instance,
+                str::from_utf8(bytes.as_ref()).unwrap().to_string(),
+            ),
+            Err(e) => (instance, e.to_string()),
+        })
+        .collect()
+        .await
+}
+
+type Responses = HashMap<String, HashMap<String, Value>>;
+
+fn merge_responses(json_resonses: Vec<(Instance, String)>) -> Responses {
+    let mut result: Responses = HashMap::new();
+
+    json_resonses
+        .into_iter()
+        .for_each(|(Instance(net, subnet), value)| {
+            let kv_subnet = result.entry(net).or_insert_with(HashMap::new);
+            kv_subnet.insert(
+                subnet,
+                serde_json::from_str(&value).unwrap_or_else(|e| Value::String(e.to_string())),
+            );
+        });
+
+    result
+}
+
+fn enrich_apis(query: &str, apis_endpoints: &mut APIsEndpoints) {
+    apis_endpoints
+        .apis
+        .iter_mut()
+        .for_each(|(_, url)| url.set_query(Some(query)))
+}
+
+async fn handle_default_request(query: &str, mut apis_endpoints: APIsEndpoints) -> Json<Responses> {
+    enrich_apis(query, &mut apis_endpoints);
+    let responses = make_requests(apis_endpoints).await;
```

Comment:
Why not to pass query into `make_requests` and call `set_query` before making a call?